File: tf_adapter/python/npu_bridge/estimator/npu_ops.py
# Copyright 2016 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Ops for collective operations implemented using hccl."""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import threading
import os
import tensorflow as tf

import numbers
from tensorflow.python.ops import array_ops
from tensorflow.python.framework import tensor_shape
from tensorflow.python.framework import dtypes
from tensorflow.python.framework import ops
# from tensorflow.contrib.offline_train.ops import gen_npu_ops
from tensorflow.contrib.util import loader
from tensorflow.python.eager import context
from tensorflow.python.framework import device
from tensorflow.python.framework import ops
from tensorflow.python.platform import resource_loader
from npu_bridge.estimator.npu.npu_common import NPUBasics

from npu_bridge.helper import helper
gen_npu_ops = helper.get_gen_ops()

DEFAULT_GRAPH_SEED = 87654321
_MAXINT32 = 2**31 - 1

def npu_test():
    """A placeholder op for values fed into the TPU simultaneously as a tuple.

    """

    return gen_npu_ops.npu_test()

def NPUInit(name=None):
  if context.executing_eagerly():
    raise RuntimeError("tf.NPUInit() is not compatible with "
                       "eager execution.")

  return gen_npu_ops.npu_init(name=name)

def NPUShutdown(name=None):
  """Shutdown a distributed NPU system for use with TensorFlow.

  Args:
      name: Name of ops.

  Returns:
      The npu init ops which will open the NPU system using `Session.run`.
  """
  if context.executing_eagerly():
    raise RuntimeError("tf.NPUShutdown() is not compatible with "
                       "eager execution.")

  return gen_npu_ops.npu_shutdown(name=name)

def initialize_system(name = None):
    """Initializes a distributed NPU system for use with TensorFlow.

    Args:
        name: Name of ops.

    Returns:
        The npu init ops which will open the NPU system using `Session.run`.
    """
    return NPUInit(name);


def shutdown_system(name = None):
    """Shuts down a running NPU system."""

    return NPUShutdown(name)


def LARS(inputs_w, inputs_g, weight_decay, hyperpara=0.001, epsilon=0.00001, name=None):
  if context.executing_eagerly():
    raise RuntimeError("tf.LARS() is not compatible with "
                       "eager execution.")

  return gen_npu_ops.lars(inputs_w=inputs_w, inputs_g=inputs_g, weight_decay=weight_decay, hyperpara=hyperpara, epsilon=epsilon, name=name)

def LARSV2(input_weight,
           input_grad,
           weight_decay,
           learning_rate,
           hyperpara=0.001,
           epsilon=0.00001,
           use_clip=False,
           name=None):
    if context.executing_eagerly():
        raise RuntimeError("tf.LARSV2() is not compatible with "
                           "eager execution.")

    return gen_npu_ops.lars_v2(input_weight=input_weight,
                               input_grad=input_grad,
                               weight_decay=weight_decay,
                               learning_rate=learning_rate,
                               hyperpara=hyperpara,
                               epsilon=epsilon,
                               use_clip=use_clip,
                               name=name)

def outfeed_dequeue_op(channel_name, output_types, output_shapes, name=None):
  return gen_npu_ops.outfeed_dequeue_op(channel_name=channel_name, output_types=output_types, output_shapes=output_shapes, name=name)

def outfeed_enqueue_op(channel_name, inputs, name=None):
  return gen_npu_ops.outfeed_enqueue_op(inputs=inputs, channel_name=channel_name, name=name)

def stop_outfeed_dequeue_op(channel_name, name=None):
  return gen_npu_ops.stop_outfeed_dequeue_op(channel_name, name)

def _truncate_seed(seed):
  return seed % _MAXINT32  # Truncate to fit into 32-bit integer

def get_seed(op_seed):
  global_seed = ops.get_default_graph().seed

  if global_seed is not None:
    if op_seed is None:
      # pylint: disable=protected-access
      op_seed = ops.get_default_graph()._last_id

    seeds = _truncate_seed(global_seed), _truncate_seed(op_seed)
  else:
    if op_seed is not None:
      seeds = DEFAULT_GRAPH_SEED, _truncate_seed(op_seed)
    else:
      seeds = None, None
  # Avoid (0, 0) as the C++ ops interpret it as nondeterminism, which would
  # be unexpected since Python docs say nondeterminism is (None, None).
  if seeds == (0, 0):
    return (0, _MAXINT32)
  return seeds

def _get_noise_shape(x, noise_shape):
  # If noise_shape is none return immediately.
  if noise_shape is None:
    return array_ops.shape(x)

  try:
    # Best effort to figure out the intended shape.
    # If not possible, let the op to handle it.
    # In eager mode exception will show up.
    noise_shape_ = tensor_shape.as_shape(noise_shape)
  except (TypeError, ValueError):
    return noise_shape

  if x.shape.dims is not None and len(x.shape.dims) == len(noise_shape_.dims):
    new_dims = []
    for i, dim in enumerate(x.shape.dims):
      if noise_shape_.dims[i].value is None and dim.value is not None:
        new_dims.append(dim.value)
      else:
        new_dims.append(noise_shape_.dims[i].value)
    return tensor_shape.TensorShape(new_dims)

  return noise_shape

def dropout(x, keep_prob, noise_shape=None, seed=None, name=None):
    """The gradient for `gelu`.

    Args:
        x: A tensor with type is float.
        keep_prob: A tensor, float, rate of every element reserved.
        noise_shape: A 1-D tensor, with type int32, shape of keep/drop what random
            generated.
        seed: Random seed.
        name: Layer name.

    Returns:
        A tensor.
    """
    if context.executing_eagerly():
      raise RuntimeError("tf.dropout() is not compatible with "
                        "eager execution.")
    x = ops.convert_to_tensor(x, name="x")
    if not x.dtype.is_floating:
      raise ValueError("x has to be a floating point tensor since it's going to"
                       " be scaled. Got a %s tensor instead." % x.dtype)
    if isinstance(keep_prob, numbers.Real) and not 0 < keep_prob <= 1:
      raise ValueError("keep_prob must be a scalar tensor or a float in the "
                       "range (0, 1], got %g" % keep_prob)
    if isinstance(keep_prob, float) and keep_prob == 1:
      return x
    seed, seed2 = get_seed(seed)
    noise_shape = _get_noise_shape(x, noise_shape)
    gen_out = gen_npu_ops.drop_out_gen_mask(noise_shape, keep_prob, seed, seed2, name)
    result = gen_npu_ops.drop_out_do_mask(x, gen_out, keep_prob, name)
    return result

@ops.RegisterGradient("DropOutDoMask")
def _DropOutDoMaskGrad(op, grad):
    result = gen_npu_ops.drop_out_do_mask(grad, op.inputs[1],  op.inputs[2])
    return [result, None, None]

def basic_lstm_cell(x, h, c, w, b, keep_prob, forget_bias, state_is_tuple,
                    activation, name=None):
    if context.executing_eagerly():
      raise RuntimeError("tf.basic_lstm_cell() is not compatible with "
                        "eager execution.")
    x = ops.convert_to_tensor(x, name="x")
    h = ops.convert_to_tensor(h, name="h")
    c = ops.convert_to_tensor(c, name="c")
    w = ops.convert_to_tensor(w, name="w")
    b = ops.convert_to_tensor(b, name="b")
    result = gen_npu_ops.basic_lstm_cell(x, h, c, w, b, keep_prob, forget_bias, state_is_tuple,
                    activation, name)
    return result

@ops.RegisterGradient("BasicLSTMCell")
def basic_lstm_cell_grad(op, dct, dht, dit, djt, dft, dot, dtanhct):

    dgate, dct_1 = gen_npu_ops.basic_lstm_cell_c_state_grad(op.inputs[2], dht, dct, op.outputs[2], op.outputs[3], op.outputs[4], op.outputs[5], op.outputs[6], forget_bias=op.get_attr("forget_bias"), activation=op.get_attr("activation"))
    dw, db = gen_npu_ops.basic_lstm_cell_weight_grad(op.inputs[0], op.inputs[1], dgate)
    dxt, dht = gen_npu_ops.basic_lstm_cell_input_grad(dgate, op.inputs[3], keep_prob=op.get_attr("keep_prob"))

    return [dxt, dht, dct_1, dw, db]

def adam_apply_one_assign(input0, input1, input2, input3, input4,
                   mul0_x, mul1_x, mul2_x, mul3_x, add2_y, name=None):
    if context.executing_eagerly():
      raise RuntimeError("tf.adam_apply_one_assign() is not compatible with "
                        "eager execution.")
    result = gen_npu_ops.adam_apply_one_assign(input0, input1, input2, input3, input4,
                   mul0_x, mul1_x, mul2_x, mul3_x, add2_y,name)
    return result

def adam_apply_one_with_decay_assign(input0, input1, input2, input3, input4,
                   mul0_x, mul1_x, mul2_x, mul3_x, mul4_x, add2_y, name=None):
    if context.executing_eagerly():
      raise RuntimeError("tf.adam_apply_one_with_decay_assign() is not compatible with "
                        "eager execution.")
    result = gen_npu_ops.adam_apply_one_with_decay_assign(input0, input1, input2, input3, input4,
                   mul0_x, mul1_x, mul2_x, mul3_x, mul4_x, add2_y, name)
    return result

File: tf_adapter/ops/npu_ops.cc
/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Copyright (C) 2019-2020. Huawei Technologies Co., Ltd. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/core/framework/common_shape_fns.h"
#include "tensorflow/core/framework/op.h"

namespace tensorflow {
using shape_inference::DimensionHandle;
using shape_inference::InferenceContext;
using shape_inference::ShapeHandle;
using shape_inference::UnchangedShape;

namespace {
REGISTER_OP("NPUTest").SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("GeOp")
    .Input("inputs: Tin")
    .Attr("Tin: list(type) >= 0")
    .Output("outputs: Tout")
    .Attr("Tout: list(type) >= 0")
    .Attr("function: func")
    .Attr("data_format: { 'NHWC', 'NCHW', 'NDHWC', 'NCDHW', 'DHWCN', 'DHWNC'} = 'NHWC'")
    .SetIsStateful();

REGISTER_OP("DPOP")
    .Input("inputs: Tin")
    .Attr("Tin: list(type) >= 0")
    .Output("outputs: Tout")
    .Attr("Tout: list(type) >= 0")
    .Attr("function: func")
    .Attr("data_format: { 'NHWC', 'NCHW'} = 'NHWC'")
    .SetIsStateful();

REGISTER_OP("NPUInit").SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("LogTimeStamp").Attr("logid: string").Attr("notify: bool").SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("NPUShutdown").SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("LARS")
    .Input("inputs_w: T")
    .Input("inputs_g: T")
    .Input("weight_decay: float")
    .Output("outputs: T")
    .Attr("T: list(type) >= 1")
    .Attr("hyperpara: float = 0.001")
    .Attr("epsilon: float = 0.00001")
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      for (int i = 0; i < ((c->num_inputs() - 1) / 2); i++) { c->set_output(i, c->input(i)); }
      return Status::OK();
    })
    .Doc(R"doc(
    Perform Lars on multi tensors. inputs_g have the same shape as `inputs_w`.

    Arguments
        inputs_w:    Tensors of weight.
        inputs_g:    Tensors of gradient.

    Output
        outputs:    Tensors with the same shape as `inputs_w`.
    )doc");

REGISTER_OP("LarsV2")
    .Input("input_weight: T")
    .Input("input_grad: T")
    .Input("weight_decay: T")
    .Input("learning_rate: T")
    .Output("output: T")
    .Attr("T: {float}")
    .Attr("hyperpara: float = 0.001")
    .Attr("epsilon: float = 0.00001")
    .Attr("use_clip: bool = false")
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      c->set_output(0, c->input(0));
      return Status::OK();
    })
    .Doc(R"doc(
    Perform LarsV2 on single output. input_weight have the same shape
    as `input_grad`.

    Arguments
        input_weight:    Tensor of weight.
        input_grad:      Tensor of gradient.
        weight_decay:    Tensor of weight_decay.
        learning_rate:   Tensor of learning_rate.
        use_clip:        Indicates whether to limit the coeff to acertain range.

    Output
        output:    Tensor with the same shape as `input_weight`.
    )doc");

Status OutfeedDequeueShapeFn(shape_inference::InferenceContext *c) {
  std::vector<PartialTensorShape> output_shapes;
  TF_RETURN_IF_ERROR(c->GetAttr("output_shapes", &output_shapes));
  if (static_cast<int>(output_shapes.size()) != c->num_outputs()) {
    return errors::InvalidArgument("`output_shapes` must be the same length as `output_types` (", output_shapes.size(),
                                   " vs. ", c->num_outputs());
  }
  for (size_t i = 0; i < output_shapes.size(); ++i) {
    shape_inference::ShapeHandle output_shape_handle;
    TF_RETURN_IF_ERROR(c->MakeShapeFromPartialTensorShape(output_shapes[i], &output_shape_handle));
    c->set_output(static_cast<int>(i), output_shape_handle);
  }
  return Status::OK();
}

REGISTER_OP("OutfeedEnqueueOp")
    .Input("inputs: Tin")
    .Attr("channel_name: string")
    .Attr("Tin: list(type) >= 0")
    .SetIsStateful()
    .SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("OutfeedDequeueOp")
    .Output("outputs: output_types")
    .Attr("channel_name: string")
    .Attr("output_types: list(type) >= 1")
    .Attr("output_shapes: list(shape) >= 1")
    .SetIsStateful()
    .SetShapeFn(OutfeedDequeueShapeFn);

REGISTER_OP("StopOutfeedDequeueOp").Attr("channel_name: string").SetIsStateful().SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("DropOutDoMask")
    .Input("x: T")
    .Input("mask: uint8")
    .Input("keep_prob: T")
    .Output("y: T")
    .Attr("T: {float16, float32}")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      c->set_output(0, c->input(0));
      return Status::OK();
    });

REGISTER_OP("DropOutGenMask")
    .Input("shape: T")
    .Attr("T: {int64, int32}")
    .Input("prob: S")
    .Attr("S: {float, half}")
    .Output("output: uint8")
    .Attr("seed: int = 0")
    .Attr("seed2: int = 0")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      ShapeHandle unused;
      TF_RETURN_IF_ERROR(c->WithRankAtMost(c->input(1), 0, &unused));  // prob must be 0-d

      ShapeHandle inputShapeHandle;
      TF_RETURN_IF_ERROR(c->MakeShapeFromShapeTensor(0, &inputShapeHandle));

      int32 rank = c->Rank(inputShapeHandle);
      if (InferenceContext::kUnknownRank == rank) {
        ShapeHandle out = c->UnknownShapeOfRank(1);
        c->set_output(0, out);
        return Status::OK();
      }

      bool unknownDimExist = false;
      for (int32 i = 0; i < rank; ++i) {
        DimensionHandle dimHandle = c->Dim(inputShapeHandle, i);
        int64 value = c->Value(dimHandle);
        if (InferenceContext::kUnknownDim == value) {
          unknownDimExist = true;
          break;
        }
      }

      if (unknownDimExist) {
        ShapeHandle out = c->UnknownShapeOfRank(1);
        c->set_output(0, out);
        return Status::OK();
      }

      int64 bitCount = 0;
      if (rank != 0) {
        DimensionHandle inputDimHandle = c->NumElements(inputShapeHandle);
        bitCount = c->Value(inputDimHandle);
      }

      // align to 128 and around up
      int64 n128Bits = bitCount / 128;
      if ((bitCount % 128) != 0) { n128Bits++; }

      // transfer 128 bit count to byte count if shape is full know
      int64 nBytes = n128Bits * 16;

      ShapeHandle out = c->Vector(nBytes);
      c->set_output(0, out);
      return Status::OK();
    });

REGISTER_OP("BasicLSTMCell")
    .Input("x: T")
    .Input("h: T")
    .Input("c: T")
    .Input("w: T")
    .Input("b: T")
    .Output("ct: T")
    .Output("ht: T")
    .Output("it: T")
    .Output("jt: T")
    .Output("ft: T")
    .Output("ot: T")
    .Output("tanhct: T")
    .Attr("T: {float16, float32}")
    .Attr("keep_prob: float = 1.0")
    .Attr("forget_bias: float = 1.0")
    .Attr("state_is_tuple: bool = true")
    .Attr("activation: string = 'tanh'")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      c->set_output(0, c->input(2));
      c->set_output(1, c->input(1));
      c->set_output(2, c->input(2));
      c->set_output(3, c->input(2));
      c->set_output(4, c->input(2));
      c->set_output(5, c->input(2));
      c->set_output(6, c->input(2));
      return Status::OK();
    });

REGISTER_OP("BasicLSTMCellCStateGrad")
    .Input("c: T")
    .Input("dht: T")
    .Input("dct: T")
    .Input("it: T")
    .Input("jt: T")
    .Input("ft: T")
    .Input("ot: T")
    .Input("tanhct: T")
    .Output("dgate: T")
    .Output("dct_1: T")
    .Attr("T: {float16, float32}")
    .Attr("forget_bias: float = 1.0")
    .Attr("activation: string = 'tanh'")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      auto input_it_shape = c->input(4);
      auto hidden_size = c->Dim(input_it_shape, 1);
      auto batch_size = c->Dim(input_it_shape, 0);
      DimensionHandle output_size;
      TF_RETURN_IF_ERROR(c->Multiply(hidden_size, 4, &output_size));
      auto output_shape = c->MakeShape({batch_size, output_size});
      c->set_output(0, output_shape);
      c->set_output(1, c->input(2));
      return Status::OK();
    });

REGISTER_OP("BasicLSTMCellWeightGrad")
    .Input("x: T")
    .Input("h: T")
    .Input("dgate: T")
    .Output("dw: T")
    .Output("db: T")
    .Attr("T: {float16, float32}")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      auto input_x_shape = c->input(0);
      auto input_h_shape = c->input(1);
      auto input_dgate_shape = c->input(2);
      auto four_hidden_size = c->Dim(input_dgate_shape, 1);
      auto hidden_size = c->Dim(input_h_shape, 1);
      auto input_size = c->Dim(input_x_shape, 1);
      DimensionHandle output_size;
      TF_RETURN_IF_ERROR(c->Add(hidden_size, input_size, &output_size));
      auto output_dw_shape = c->MakeShape({output_size, four_hidden_size});
      auto output_db_shape = c->MakeShape({four_hidden_size});
      c->set_output(0, output_dw_shape);
      c->set_output(1, output_db_shape);
      return Status::OK();
    });

REGISTER_OP("BasicLSTMCellInputGrad")
    .Input("dgate: T")
    .Input("w: T")
    .Output("dxt: T")
    .Output("dht: T")
    .Attr("T: {float16, float32}")
    .Attr("keep_prob: float = 1.0")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      auto input_dgate_shape = c->input(0);
      auto input_w_shape = c->input(1);
      auto four_hidden_size = c->Dim(input_dgate_shape, 1);
      auto batch_size = c->Dim(input_dgate_shape, 0);
      auto input_hidden_size = c->Dim(input_w_shape, 0);
      DimensionHandle output_hidden_size;
      TF_RETURN_IF_ERROR(c->Divide(four_hidden_size, 4, true, &output_hidden_size));
      auto output_dht_shape = c->MakeShape({batch_size, output_hidden_size});
      DimensionHandle output_input_size;
      TF_RETURN_IF_ERROR(c->Subtract(input_hidden_size, output_hidden_size, &output_input_size));
      auto output_dxt_shape = c->MakeShape({batch_size, output_input_size});
      c->set_output(0, output_dxt_shape);
      c->set_output(1, output_dht_shape);
      return Status::OK();
    });

REGISTER_OP("DecodeAndResizeJpeg")
    .Input("contents: string")
    .Input("resize: int32")
    .Output("image: uint8")
    .SetShapeFn([](InferenceContext *c) {
      const Tensor *rezise = c->input_tensor(1);
      DimensionHandle h;
      DimensionHandle w;
      if (rezise != nullptr) {
        auto rezise_vec = rezise->vec<int32>();
        h = c->MakeDim(rezise_vec(0));
        w = c->MakeDim(rezise_vec(1));
      }
      c->set_output(0, c->MakeShape({h, w, 3}));
      return Status::OK();
    });

REGISTER_OP("DecodeAndCropAndResizeJpeg")
    .Input("contents: string")
    .Input("crop_size: int32")
    .Input("resize: int32")
    .Output("image: uint8")
    .SetShapeFn([](InferenceContext *c) {
      const Tensor *rezise = c->input_tensor(2);
      DimensionHandle h;
      DimensionHandle w;
      if (rezise != nullptr) {
        auto rezise_vec = rezise->vec<int32>();
        h = c->MakeDim(rezise_vec(0));
        w = c->MakeDim(rezise_vec(1));
      }
      c->set_output(0, c->MakeShape({h, w, 3}));
      return Status::OK();
    });

REGISTER_OP("AdamApplyOneAssign")
    .Input("input0: T")
    .Input("input1: T")
    .Input("input2: T")
    .Input("input3: T")
    .Input("input4: T")
    .Input("mul0_x: T")
    .Input("mul1_x: T")
    .Input("mul2_x: T")
    .Input("mul3_x: T")
    .Input("add2_y: T")
    .Attr("T: {float16, float32}")
    .SetShapeFn(shape_inference::NoOutputs);

REGISTER_OP("LambApplyOptimizerAssign")
    .Input("input0: T")
    .Input("input1: T")
    .Input("input2: T")
    .Input("input3: T")
    .Input("mul0_x: T")
    .Input("mul1_x: T")
    .Input("mul2_x: T")
    .Input("mul3_x: T")
    .Input("add2_y: T")
    .Input("steps: T")
    .Input("do_use_weight: T")
    .Input("weight_decay_rate: T")
    .Output("update: T")
    .Output("output1: T")
    .Output("output2: T")
    .Attr("T: {float16, float32}")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      c->set_output(0, c->input(0));
      c->set_output(1, c->input(1));
      c->set_output(2, c->input(2));
      return Status::OK();
    });

REGISTER_OP("LambApplyWeightAssign")
    .Input("input0: T")
    .Input("input1: T")
    .Input("input2: T")
    .Input("input3: T")
    .Input("input4: T")
    .Output("output0: T")
    .Attr("T: {float16, float32}")
    .SetIsStateful()
    .SetShapeFn([](shape_inference::InferenceContext *c) {
      c->set_output(0, c->input(4));
      return Status::OK();
    });

REGISTER_OP("AdamApplyOneWithDecayAssign")
    .Input("input0: T")
    .Input("input1: T")
    .Input("input2: T")
    .Input("input3: T")
    .Input("input4: T")
    .Input("mul0_x: T")
    .Input("mul1_x: T")
    .Input("mul2_x: T")
    .Input("mul3_x: T")
    .Input("mul4_x: T")
    .Input("add2_y: T")
    .Attr("T: {float16, float32}")
    .SetShapeFn(shape_inference::NoOutputs);
}  // namespace
}  // namespace tensorflow
